`science` blows up when `~/.netrc` is a directory.

It's true it's strange to have a `~/.netrc` directory, but GitHub actions Linux runners appear to (at least ubuntu-22.04. As seen over in this [Pex PR](https://github.com/pex-tool/pex/pull/2466) when trying to use science in CI:
https://github.com/pex-tool/pex/actions/runs/9961092522/job/27521849245?pr=2466#step:5:3152

That's:
```
E       pex: Building scie(s): 4215.2ms
E       Traceback (most recent call last):
E         File "/development/pex/pex/result.py", line 105, in catch
E           return func(*args, **kwargs)
E                  ^^^^^^^^^^^^^^^^^^^^^
E         File "/development/pex/pex/bin/pex.py", line 1311, in do_main
E           for par_info in scie.build(
E         File "/development/pex/pex/scie/science.py", line 355, in build
E           raise ScienceError(
E       pex.scie.science.ScienceError: Failed to build 1 scie:
E       
E       1. For CPython 3.11 on linux-x86_64: Command `/home/runner/.pex/scies/science/0.3.0/bin/science --cache-dir /home/runner/.pex/scies/science/0.3.0/cache -vvvvv lift --file pex=/tmp/pytest-of-runner/pytest-0/popen-gw1/test_basic_ZIPAPP_packed_eager0/cowsay.pex --file configure-binding.py=/development/pex/pex/scie/configure-binding.py build --dest-dir /tmp/pytest-of-runner/pytest-0/popen-gw1/test_basic_ZIPAPP_packed_eager0 ../../tmp/pytest-of-runner/pytest-0/popen-gw1/test_basic_ZIPAPP_packed_eager0/cowsay-lift-linux-x86_64.toml` failed with exit code 1 (saved lift manifest to ../../tmp/pytest-of-runner/pytest-0/popen-gw1/test_basic_ZIPAPP_packed_eager0/cowsay-lift-linux-x86_64.toml for inspection):
E       debug: Attempting to acquire lock 140188457509264 on /home/runner/.pex/scies/science/0.3.0/cache/downloads/1a5963796446dc5314479d9b05931d6a79b2ecdb7c599cc59cfa46416b2e0bf1.lck
E       debug: Lock 140188457509264 acquired on /home/runner/.pex/scies/science/0.3.0/cache/downloads/1a5963796446dc5314479d9b05931d6a79b2ecdb7c599cc59cfa46416b2e0bf1.lck
E       debug: Attempting to release lock 140188457509264 on /home/runner/.pex/scies/science/0.3.0/cache/downloads/1a5963796446dc5314479d9b05931d6a79b2ecdb7c599cc59cfa46416b2e0bf1.lck
E       debug: Lock 140188457509264 released on /home/runner/.pex/scies/science/0.3.0/cache/downloads/1a5963796446dc5314479d9b05931d6a79b2ecdb7c599cc59cfa46416b2e0bf1.lck
E         File "<frozen runpy>", line 198, in _run_module_as_main
E         File "<frozen runpy>", line 88, in _run_code
E         File "/home/runner/.cache/nce/21383839c05483df915abdaff3ebfbbe4bd7bb2b2543135d616aa7278db327f3/science.pyz/__main__.py", line 3, in <module>
E           _bootstrap.bootstrap()
E         File "/home/runner/.cache/nce/21383839c05483df915abdaff3ebfbbe4bd7bb2b2543135d616aa7278db327f3/science.pyz/_bootstrap/__init__.py", line 253, in bootstrap
E           run(import_string(env.entry_point))
E         File "/home/runner/.cache/nce/21383839c05483df915abdaff3ebfbbe4bd7bb2b2543135d616aa7278db327f3/science.pyz/_bootstrap/__init__.py", line 38, in run
E           sys.exit(module())
E                    ^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/science/exe.py", line 692, in main
E           _main(prog_name=SCIE_ARGV0)
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/click/core.py", line 1130, in __call__
E           return self.main(*args, **kwargs)
E                  ^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/click/core.py", line 1055, in main
E           rv = self.invoke(ctx)
E                ^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/click/core.py", line 1657, in invoke
E           return _process_result(sub_ctx.command.invoke(sub_ctx))
E                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/click/core.py", line 1657, in invoke
E           return _process_result(sub_ctx.command.invoke(sub_ctx))
E                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/click/core.py", line 1404, in invoke
E           return ctx.invoke(self.callback, **ctx.params)
E                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/click/core.py", line 760, in invoke
E           return __callback(*args, **kwargs)
E                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/click/decorators.py", line 84, in new_func
E           return ctx.invoke(f, obj, *args, **kwargs)
E                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/click/core.py", line 760, in invoke
E           return __callback(*args, **kwargs)
E                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/science/exe.py", line 633, in _build
E           application = parse_application(lift_config, config)
E                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/science/exe.py", line 518, in parse_application
E           application = parse_config(config, source=config.name)
E                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/science/config.py", line 41, in parse_config
E           return parse_config_data(Data(provenance=provenance, data=data))
E                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/science/config.py", line 141, in parse_config_data
E           interpreters_by_id = {
E                                ^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/science/config.py", line 143, in <dictcomp>
E           interp := parse_dataclass(
E                     ^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/science/dataclass/deserializer.py", line 177, in parse
E           kwargs[field.name] = parser(data)
E                                ^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/science/config.py", line 72, in parse_provider
E           return provider_type.create(identifier=fields.id, lazy=fields.lazy, config=config)
E                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/science/providers/python_build_standalone.py", line 176, in create
E           release_data = fetch_json(release_url, ttl=ttl)
E                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/science/fetcher.py", line 115, in fetch_json
E           with _fetch_to_cache(url, ttl, headers).open() as fp:
E                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/science/fetcher.py", line 98, in _fetch_to_cache
E           with _configured_client(url, headers).stream("GET", url) as response, work.open(
E                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/science/fetcher.py", line 88, in _configured_client
E           auth = _configure_auth(url) if "Authorization" not in headers else None
E                  ^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/science/fetcher.py", line 77, in _configure_auth
E           return httpx.NetRCAuth(None)
E                  ^^^^^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/1c179c70d114c64117246043677b89be985cdfd87bf4c3b3aea0844cd78c49bc/bindings/shiv_root/science.pyz_154ac1f5c68d25395a1d3d135b4392c3a8ba0b549c7c6fb7a9376dd3d0a34a80/site-packages/httpx/_auth.py", line 151, in __init__
E           self._netrc_info = netrc.netrc(file)
E                              ^^^^^^^^^^^^^^^^^
E         File "/home/runner/.cache/nce/fbed6f7694b2faae5d7c401a856219c945397f772eea5ca50c6eb825cbc9d1e1/cpython-3.11.5+20230826-x86_64-unknown-linux-gnu-install_only.tar.gz/python/lib/python3.11/netrc.py", line 74, in __init__
E           with open(file, encoding="utf-8") as fp:
E                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       builtins.IsADirectoryError: [Errno 21] Is a directory: '/home/runner/.netrc'
```

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,10 @@
 # Release Notes
 
+## 0.4.3
+
+Fix science URL fetching code to gracefully ignore a `~/.netrc` that is a directory when configuring
+authentication for the fetch.
+
 ## 0.4.2
 
 Upgrade the science internal Python distribution to [PBS][PBS] CPython 3.12.4.

--- a/science/__init__.py
+++ b/science/__init__.py
@@ -3,6 +3,6 @@
 
 from packaging.version import Version
 
-__version__ = "0.4.2"
+__version__ = "0.4.3"
 
 VERSION = Version(__version__)

--- a/science/fetcher.py
+++ b/science/fetcher.py
@@ -75,7 +75,7 @@ def _configure_auth(url: Url) -> httpx.Auth | tuple[str, str] | None:
 
     try:
         return httpx.NetRCAuth(None)
-    except FileNotFoundError:
+    except (FileNotFoundError, IsADirectoryError):
         pass
     except NetrcParseError as e:
         logger.warning(f"Not using netrc for auth, netrc file is invalid: {e}")

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -4,8 +4,13 @@
 import os
 import sys
 from pathlib import Path
+from typing import Iterator
 
 import pytest
+from click import Command, Context
+from click.globals import pop_context, push_context
+
+from science.context import ScienceConfig
 
 
 def pytest_sessionstart(session: pytest.Session) -> None:
@@ -26,3 +31,13 @@ def science_pyz() -> Path:
         return Path(os.environ["SCIENCE_TEST_PYZ_PATH"])
     except KeyError:
         pytest.fail("Test must be run via `nox` or `nox -etest`.")
+
+
+@pytest.fixture
+def cache_dir(tmp_path: Path) -> Iterator[Path]:
+    cache_dir = tmp_path / "nce"
+    push_context(Context(Command(None), obj=ScienceConfig(cache_dir=cache_dir)))
+    try:
+        yield cache_dir
+    finally:
+        pop_context()

--- a/tests/test_fetcher.py
+++ b/tests/test_fetcher.py
@@ -1,0 +1,34 @@
+# Copyright 2024 Science project contributors.
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import shutil
+from pathlib import Path
+
+from pytest import MonkeyPatch
+from testing import issue
+
+from science.fetcher import fetch_text
+from science.model import Url
+
+
+@issue(71, ignore=True)
+def test_netrc_directory(_, tmp_path: Path, monkeypatch: MonkeyPatch, cache_dir: Path) -> None:
+    def assert_fetch() -> None:
+        # N.B.: Ensure a fresh, un-cached fetch.
+        shutil.rmtree(cache_dir, ignore_errors=True)
+        assert (
+            "7ed49bb4c50960d4ade3cf9a5614bd9c1190cc57d330492e36a7ace22b8ebc3b "
+            "*science-fat-linux-aarch64\n"
+        ) == fetch_text(
+            Url(
+                "https://github.com/a-scie/lift/releases/download/v0.4.2/"
+                "science-fat-linux-aarch64.sha256"
+            )
+        )
+
+    assert_fetch()
+
+    home_dir = tmp_path / "home"
+    monkeypatch.setenv("HOME", str(home_dir))
+    (home_dir / ".netrc").mkdir(parents=True)
+    assert_fetch()

